src/bin/shard/buddy.rs

```diff
@@ -340,3 +341,85 @@ pub fn find_name_accept_buddy(
         },
     )
 }
+
+pub fn buddy_warp(clients: &mut ClientMap, state: &mut ShardServerState) -> FFResult<()> {
+    let client = clients.get_self();
+    let pkt: sP_CL2FE_REQ_PC_BUDDY_WARP = *client.get_packet(P_CL2FE_REQ_PC_BUDDY_WARP)?;
+
+    let pc_id = client.get_player_id()?;
+    let player = state.get_player(pc_id)?;
+
+    let buddy_uid = pkt.iBuddyPCUID;
+    if !player.is_buddies_with(buddy_uid) {
+        return Err(FFError::build(
+            Severity::Warning,
+            format!("{} is not buddies with player {}", player, buddy_uid),
+        ));
+    }
+
+    let search = PlayerSearchQuery::ByUID(buddy_uid);
+    let res = search.execute(state);
+    if res.is_none() {
+        if let Some(login_server) = clients.get_login_server() {
+            let cross_shard_pkt = sP_FE2LS_REQ_BUDDY_WARP {
+                iPC_ID: pc_id,
+                iPCUID: player.get_uid(),
+                iBuddyPCUID: buddy_uid,
+            };
+            log_if_failed(login_server.send_packet(P_FE2LS_REQ_BUDDY_WARP, &cross_shard_pkt));
+        } else {
+            return Err(FFError::build(
+                Severity::Warning,
+                "No login server connected for cross-shard buddy warp".to_string(),
+            ));
+        }
+        return Ok(());
+    }
+
+    let buddy_id = res.unwrap();
+    let buddy: &rusty_fusion::entity::Player = state.get_player(buddy_id)?;
+
+    if buddy.instance_id.map_num != 0 {
+        let response = sP_FE2CL_REP_PC_BUDDY_WARP_FAIL {
+            iBuddyPCUID: buddy_uid,
+            iErrorCode: 1, // TODO: define error codes
+        };
+
+        log_if_failed(
+            clients
+                .get_self()
+                .send_packet(P_FE2CL_REP_PC_BUDDY_WARP_FAIL, &response),
+        );
+
+        return Err(FFError::build(
+            Severity::Warning,
+            format!(
+                "Cannot warp to buddy {} - buddy is inside an instance zone (map_num: {})",
+                buddy_uid, buddy.instance_id.map_num
+            ),
+        ));
+    }
+
+    let buddy_pos = buddy.get_position();
+
+    let player = state.get_player_mut(pc_id)?;
+    player.set_position(buddy_pos);
+    player.instance_id = InstanceID::default();
+
+    state
+        .entity_map
+        .update(EntityID::Player(pc_id), None, Some(clients));
+
+    let resp = sP_FE2CL_REP_PC_GOTO_SUCC {
+        iX: buddy_pos.x,
+        iY: buddy_pos.y,
+        iZ: buddy_pos.z,
+    };
+    log_if_failed(
+        clients
+            .get_self()
+            .send_packet(P_FE2CL_REP_PC_GOTO_SUCC, &resp),
```

Comment:
this probably works fine but I'm wondering if `P_FE2CL_REP_PC_BUDDY_WARP_SAME_SHARD_SUCC` is relevant to us at all